### ISSUE TITLE
Expired certificates cannot be revoked

### ISSUE THREAD

**Issue by Frizlab** [hashicorp/vault#19452]
**Describe the bug**
It is not possible to revoke expired certificates.

**To Reproduce**
1. Create a certificate expiring soon (e.g. in 5s);
2. Wait for the certificate to expire;
3. Revoke it;
4. We get a 204 and nothing else happens.

**Expected behavior**
The CRL should be updated with the new certificate.
At the very least an option should be provided to be able to force revoke expired certificates.

**Environment:**
* Vault Server Version: `Vault v1.7.2 (db0e4245d5119b5929e611ea4d9bf66e47f3f208)`
* Vault CLI Version: N/A (I used API calls)
* Server Operating System/Architecture: Debian x86

Vault server configuration file(s):
```hcl
ui = "true"

storage "gcs" {
  bucket = "XXX"
  ha_enabled = "true"
}

listener "tcp" {
  address     = "0.0.0.0:8200"
  cluster_addr = "https://XXX"
  tls_cert_file = "/etc/ssl/certs/XXX.pem"
  tls_key_file = "/etc/ssl/private/XXX.pem"
}

seal "gcpckms" {
  project = "XXX"
  region = "XXX"
  key_ring = "XXX"
  crypto_key = "XXX"
}

api_addr = "https://XXX"
cluster_addr = "https://XXX"
```

**Additional context**
None.

**Comment by hsimon-hashicorp** [comment]
Hello! For new bug reports and issues, we maintain a version n-2 model. 1.7.2 was released in 2021. Can you verify whether this same behavior occurs in a newer version? Thanks! https://developer.hashicorp.com/vault/downloads

**Comment by cipherboy** [comment]
@Frizlab What is the purpose of revoking expired certificates? RFC 5280 is very clear that status information only needs to be maintained until the `notAfter` date, e.g.:

 - https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.5 -- CA warrants that it will maintain information about the status of the
   certificate.  /  If a CA is unable to maintain status information _until_ a certificate's notAfter... 
 - https://datatracker.ietf.org/doc/html/rfc5280#section-5 -- A complete CRL lists all _unexpired_ certificates...

&c.

There's no mandate to process revocations after a certificate has expired, as client should still consider them untrusted except in exceptional scenarios (excessive clock drift, whether manual or intentional).

**Comment by Frizlab** [comment]
@hsimon-hashicorp Yes, I get the same behavior with `Vault v1.13.0 ('a4cf0dc4437de35fce4860857b64569d092a9b5a+CHANGES'), built 2023-03-01T14:58:13Z`.

@cipherboy I agree, there is no obligation for the CRL to send expired certificates. `vault` does do that if a certificate has been revoked before its expiration date. I see no reasons why we could not add expired certificates to the CRL.

My use case: I want to have the list of my certificates sorted by expiration date in order to be able to renew the ones expiring soon, or already expired. I do not want to see revoked certificates, and I want to be able to “remove” a certificate from the list even if it is expired, by revoking it.
I agree the use case is probably rare and I have workarounds if the team do not want to add an option to revoke expired certificates.

**Comment by cipherboy** [comment]
@Frizlab This sounds like a bit of an unexpected use case for the CRL and definitely not something encouraged by the RFC. :-)

I don't think we'll be supporting this particular usage of CRLs for this, but we're definitely interested in the renewal problem in general.
